File: brainfuck/fsharp/bf.fs
open System
module Tape =
    type t = { mutable data: int array; mutable pos: int }

    let create () = { data = [| 0 |]; pos = 0 }
    let current t = t.data.[t.pos]
    let inc t delta =
        t.data.[t.pos] <- current t + delta

    let put t ch =
        t.data.[t.pos] <- ch
        
    let move t m =
        let newPos = t.pos + m
        if newPos >= Array.length t.data then
            Array.Resize (&t.data, 2 * newPos)
        t.pos <- newPos

module Printer =
    type sum =
        { mutable sum1: int
          mutable sum2: int }
    
    [<Struct>]
    type t =
        | Quiet of sum
        | Stdout

    let createQuiet () = Quiet { sum1 = 0; sum2 = 0 }
    let createStdout () = Stdout

    let print p n =
        match p with
        | Stdout ->
            Console.Out.Write(char n)
            Console.Out.Flush()
        | Quiet p ->
            let newSum1 = (p.sum1 + n) % 255
            let newSum2 = (newSum1 + p.sum2) % 255
            p.sum1 <- newSum1
            p.sum2 <- newSum2
        p
    let checksum = function
        | Stdout -> 0
        | Quiet sm -> (sm.sum2 <<< 8) ||| sm.sum1

    let isQuiet = function
        | Quiet _ -> true
        | _ -> false

module Interpreter =
    
    [<Struct>]
    type Op =
        | Inc of delta: int
        | Move of steps: int
        | Print
        | Input
        | Loop of opcodes: List<Op>
        | Ignore

    let parse (s: string) =
        let opCodes = Set.ofSeq "+-<>,."
        let rec _parse res = function
            | [] -> List.rev res, List.empty
            | ']' :: tail -> List.rev res, tail
            | '[' :: tail ->
                let (codes, newTail) = _parse [] tail
                _parse (Loop codes :: res) newTail
            | ch :: tail when Set.contains ch opCodes  ->
                let code =
                    match ch with
                    | '+' -> Inc 1
                    | '-' -> Inc -1
                    | '>' -> Move 1
                    | '<' -> Move -1
                    | '.' -> Print
                    | ',' -> Input
                    | _ -> Ignore
                _parse (code :: res) tail
            | _ :: tail -> _parse res tail            
        Seq.toList s
            |> _parse []
            |> fst

    let rec run tape printer program = 
        let rec _run = function
            | [] -> ()
            | Inc delta :: restProgram ->
                Tape.inc tape delta
                _run  restProgram
            | Move delta :: rest ->
                Tape.move tape delta
                _run rest
            | Print :: rest ->
                Tape.current tape |> Printer.print printer |> ignore
                _run rest
            | Loop loopCode :: rest ->
                    while Tape.current tape > 0 do
                        _run loopCode
                    _run rest
            | Input :: rest ->
                let c = Console.Read()
                Tape.put tape c
                _run rest
            | Ignore :: rest -> _run rest
        _run program
    
    let runWithNewTape printer program =
        run (Tape.create ()) printer program
        printer

open System.Diagnostics
open System.IO
module Util =
    let notify (msg: string) =
        try
            use s = new System.Net.Sockets.TcpClient("localhost", 9001)
            msg
                |> System.Text.Encoding.UTF8.GetBytes
                |> s.Client.Send
                |> ignore
        with _ -> ()

    let verify () =
        let source = """
            ++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>
            ---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.
            """
        let left =
            Interpreter.parse source
            |> Interpreter.runWithNewTape (Printer.createQuiet ())
            |> Printer.checksum

        let right =
            "Hello World!\n"
            |> Seq.map int
            |> Seq.fold Printer.print (Printer.createQuiet())
            |> Printer.checksum

        if left <> right
        then Result.Error <| sprintf "%d != %d" left right 
        else Result.Ok None

    let isQuiet =
        Environment.GetEnvironmentVariable "QUIET"
        <> null

    let runtimeId =
        let runtime = if isNull (Type.GetType "Mono.Runtime") then ".NET Core" else "Mono"
        sprintf "F#/%s\t%d" runtime (Process.GetCurrentProcess().Id)

[<EntryPoint>]
let main argv =
    match Util.verify () with
    | Error error ->
        printfn "%A" error
        exit 1
    | Ok _ -> ()

    match argv with
    | [| filename |] ->
        let source =
            try
                File.ReadAllText filename
            with ex ->
                printfn "Can't open file '%O': %A" filename ex.Message
                exit 1

        let printer = if Util.isQuiet then Printer.createQuiet () else Printer.createStdout ()

        Util.notify Util.runtimeId

        let stopWatch = Stopwatch.StartNew()

        let pNew =
            Interpreter.parse source
            |> Interpreter.runWithNewTape printer

        stopWatch.Stop()

        let elapsed = stopWatch.Elapsed.TotalSeconds

        Util.notify "stop"
        printfn "time: %Os" elapsed

        if Printer.isQuiet pNew then
            Printer.checksum pNew |> printfn "Output checksum: %d"
        0
    | _ ->
        printfn "usage: bf.fs <brainfuck.b>"
        1

File: brainfuck/bf.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Test
{
    enum OpT { INC, MOVE, PRINT, LOOP };

    struct Op
    {
        public OpT op;
        public int v;
        public Op[] loop;

        public Op(OpT _op, int _v)
        {
            op = _op;
            v = _v;
            loop = null;
        }

        public Op(OpT _op, Op[] _l)
        {
            op = _op;
            loop = _l;
            v = 0;
        }
    }

    public class Tape
    {
        int pos = 0;
        int[] tape = new int[1];
        public int CurrentCell
        {
            get => tape[pos];
            set => tape[pos] = value;
        }

        public void Inc(int x) => CurrentCell += x;
        public void Move(int x)
        {
            pos += x;
            if (pos >= tape.Length)
                Array.Resize(ref tape, tape.Length * 2);
        }
    }

    class Printer
    {
        int sum1 = 0;
        int sum2 = 0;
        public int Checksum
        {
            get => (sum2 << 8) | sum1;
        }

        public bool Quiet { get; set; }

        public void Print(int n)
        {
            if (Quiet)
            {
                sum1 = (sum1 + n) % 255;
                sum2 = (sum2 + sum1) % 255;
            }
            else
            {
                Console.Write((char)n);
            }
        }
    }

    class Program
    {
        Op[] ops;
        Printer p;

        Program(string text, Printer p)
        {
            ops = parse(text.GetEnumerator());
            this.p = p;
        }

        private Op[] parse(IEnumerator<char> it)
        {
            List<Op> res = new List<Op>();
            while (it.MoveNext())
            {
                switch (it.Current)
                {
                    case '+': res.Add(new Op(OpT.INC, 1)); break;
                    case '-': res.Add(new Op(OpT.INC, -1)); break;
                    case '>': res.Add(new Op(OpT.MOVE, 1)); break;
                    case '<': res.Add(new Op(OpT.MOVE, -1)); break;
                    case '.': res.Add(new Op(OpT.PRINT, 0)); break;
                    case '[': res.Add(new Op(OpT.LOOP, parse(it))); break;
                    case ']': return res.ToArray();
                }
            }
            return res.ToArray();
        }

        public void run() => _run(ops, new Tape());

        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
        private void _run(Op[] program, Tape tape)
        {
            for (var index = 0; index < program.Length; index++)
            {
                var op = program[index];
                if (op.op == OpT.INC)
                    tape.Inc(op.v);
                else if (op.op == OpT.MOVE)
                    tape.Move(op.v);
                else if (op.op == OpT.LOOP)
                    while (tape.CurrentCell > 0)
                        _run(op.loop, tape);
                else if (op.op == OpT.PRINT) p.Print(tape.CurrentCell);
            }
        }

        private static void Notify(string msg)
        {
            try
            {
                using (var s = new System.Net.Sockets.TcpClient("localhost", 9001))
                {
                    var data = System.Text.Encoding.UTF8.GetBytes(msg);
                    s.Client.Send(data);
                }
            }
            catch
            {
                // standalone usage
            }
        }

        private static void Verify()
        {
            var text = @"++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>
                ---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.";
            var p_left = new Printer { Quiet = true };
            new Program(text, p_left).run();
            var left = p_left.Checksum;

            var p_right = new Printer { Quiet = true };
            foreach (var c in "Hello World!\n")
            {
                p_right.Print(c);
            }
            var right = p_right.Checksum;

            if (left != right)
            {
                Console.Error.WriteLine($"{left} != {right}");
                System.Environment.Exit(1);
            }
        }

        static void Main(string[] args)
        {
            Verify();
            var text = File.ReadAllText(args[0]);
            var p = new Printer
            {
                Quiet = Environment.GetEnvironmentVariable("QUIET") != null
            };

            var runtime = Type.GetType("Mono.Runtime") != null ? "Mono" : ".NET Core";
            Notify($"C#/{runtime}\t{Process.GetCurrentProcess().Id}");
            var stopWatch = Stopwatch.StartNew();

            new Program(text, p).run();
            stopWatch.Stop();
            var elapsed = stopWatch.ElapsedMilliseconds / 1e3;

            Notify("stop");
            Console.Error.WriteLine($"time: {elapsed}s");

            if (p.Quiet)
            {
                Console.WriteLine($"Output checksum: {p.Checksum}");
            }
        }
    }
}
